File: ios/Classes/AllPaystackPaymentsPlugin.swift
import Flutter
import UIKit

public class AllPaystackPaymentsPlugin: NSObject, FlutterPlugin {
  private var registrar: FlutterPluginRegistrar?
  private var paystackPublicKey: String?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "all_paystack_payments", binaryMessenger: registrar.messenger())
    let instance = AllPaystackPaymentsPlugin()
    instance.registrar = registrar
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "initialize":
      handleInitialize(call, result: result)
    case "initializePayment":
      handleInitializePayment(call, result: result)
    case "getCheckoutUrl":
      handleGetCheckoutUrl(call, result: result)
    case "verifyPayment":
      handleVerifyPayment(call, result: result)
    case "getPaymentStatus":
      handleGetPaymentStatus(call, result: result)
    case "cancelPayment":
      handleCancelPayment(call, result: result)
    case "showWebView":
      handleShowWebView(call, result: result)
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func handleInitialize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let publicKey = args["publicKey"] as? String else {
      result(FlutterError(code: "INVALID_ARGUMENTS", message: "Public key is required", details: nil))
      return
    }

    paystackPublicKey = publicKey
    result(nil)
  }

  private func handleInitializePayment(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    // For unified webview approach, initializePayment now gets checkout URL and returns it
    // The actual payment processing happens in the webview
    handleGetCheckoutUrl(call, result: result)
  }

  private func handleGetCheckoutUrl(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let amount = args["amount"] as? Int,
          let email = args["email"] as? String else {
      result(FlutterError(code: "INVALID_ARGUMENTS", message: "Amount and email are required", details: nil))
      return
    }

    // Initialize transaction with Paystack API
    initializeTransaction(amount: amount, email: email, reference: args["reference"] as? String, currency: args["currency"] as? String, metadata: args["metadata"] as? [String: Any], callbackUrl: args["callbackUrl"] as? String, result: result)
  }

  private func initializeTransaction(amount: Int, email: String, reference: String?, currency: String?, metadata: [String: Any]?, callbackUrl: String?, result: @escaping FlutterResult) {
    guard let publicKey = paystackPublicKey else {
      result(FlutterError(code: "NOT_INITIALIZED", message: "Paystack not initialized", details: nil))
      return
    }

    // Prepare request data
    var requestData: [String: Any] = [
      "amount": amount,
      "email": email
    ]

    if let reference = reference {
      requestData["reference"] = reference
    }
    if let currency = currency {
      requestData["currency"] = currency
    }
    if let metadata = metadata {
      requestData["metadata"] = metadata
    }
    if let callbackUrl = callbackUrl {
      requestData["callback_url"] = callbackUrl
    }

    // Convert to JSON
    guard let jsonData = try? JSONSerialization.data(withJSONObject: requestData, options: []) else {
      result(FlutterError(code: "SERIALIZATION_ERROR", message: "Failed to serialize request data", details: nil))
      return
    }

    // Create URL request
    guard let url = URL(string: "https://api.paystack.co/transaction/initialize") else {
      result(FlutterError(code: "INVALID_URL", message: "Invalid API URL", details: nil))
      return
    }

    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue("Bearer \(publicKey)", forHTTPHeaderField: "Authorization")
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.httpBody = jsonData

    // Make the request
    let task = URLSession.shared.dataTask(with: request) { (data, response, error) in
      if let error = error {
        result(FlutterError(code: "NETWORK_ERROR", message: "Network request failed: \(error.localizedDescription)", details: nil))
        return
      }

      guard let data = data else {
        result(FlutterError(code: "NO_DATA", message: "No data received from server", details: nil))
        return
      }

      do {
        guard let jsonResponse = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any],
              let status = jsonResponse["status"] as? Bool else {
          result(FlutterError(code: "PARSE_ERROR", message: "Invalid response format", details: nil))
          return
        }

        if !status {
          let message = jsonResponse["message"] as? String ?? "Transaction initialization failed"
          result(FlutterError(code: "API_ERROR", message: message, details: nil))
          return
        }

        guard let data = jsonResponse["data"] as? [String: Any],
              let checkoutUrl = data["authorization_url"] as? String else {
          result(FlutterError(code: "PARSE_ERROR", message: "No checkout URL in response", details: nil))
          return
        }

        // Return the checkout URL for webview processing
        let response: [String: Any] = [
          "status": "success",
          "data": [
            "authorization_url": checkoutUrl,
            "reference": data["reference"] as? String ?? ""
          ]
        ]
        result(response)
      } catch {
        result(FlutterError(code: "PARSE_ERROR", message: "Failed to parse API response: \(error.localizedDescription)", details: nil))
      }
    }
    task.resume()
  }

  private func handleVerifyPayment(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let reference = args["reference"] as? String else {
      result(FlutterError(code: "INVALID_ARGUMENTS", message: "Reference is required", details: nil))
      return
    }

    verifyTransaction(reference: reference, result: result)
  }

  private func handleGetPaymentStatus(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let reference = args["reference"] as? String else {
      result(FlutterError(code: "INVALID_ARGUMENTS", message: "Reference is required", details: nil))
      return
    }

    verifyTransaction(reference: reference, result: result)
  }

  private func handleCancelPayment(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let reference = args["reference"] as? String else {
      result(FlutterError(code: "INVALID_ARGUMENTS", message: "Reference is required", details: nil))
      return
    }

    // Paystack doesn't support canceling transactions once initiated
    result(false)
  }

  private func handleShowWebView(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
          let checkoutUrl = args["checkoutUrl"] as? String else {
      result(FlutterError(code: "INVALID_ARGUMENTS", message: "Checkout URL is required", details: nil))
      return
    }

    // For now, return a simulated response
    // In a full implementation, this would show a webview and handle callbacks
    let response: [String: Any] = [
      "status": "success",
      "message": "Payment completed",
      "data": [
        "reference": "ios_webview_ref_\(Int(Date().timeIntervalSince1970))",
        "status": "success"
      ]
    ]
    result(response)
  }

  private func verifyTransaction(reference: String, result: @escaping FlutterResult) {
    guard let publicKey = paystackPublicKey else {
      result(FlutterError(code: "NOT_INITIALIZED", message: "Paystack not initialized", details: nil))
      return
    }

    guard let url = URL(string: "https://api.paystack.co/transaction/verify/\(reference)") else {
      result(FlutterError(code: "INVALID_URL", message: "Invalid API URL", details: nil))
      return
    }

    var request = URLRequest(url: url)
    request.httpMethod = "GET"
    request.setValue("Bearer \(publicKey)", forHTTPHeaderField: "Authorization")

    let task = URLSession.shared.dataTask(with: request) { (data, response, error) in
      if let error = error {
        result(FlutterError(code: "NETWORK_ERROR", message: "Network request failed: \(error.localizedDescription)", details: nil))
        return
      }

      guard let data = data else {
        result(FlutterError(code: "NO_DATA", message: "No data received from server", details: nil))
        return
      }

      do {
        guard let jsonResponse = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] else {
          result(FlutterError(code: "PARSE_ERROR", message: "Invalid response format", details: nil))
          return
        }

        result(jsonResponse)
      } catch {
        result(FlutterError(code: "PARSE_ERROR", message: "Failed to parse API response: \(error.localizedDescription)", details: nil))
      }
    }
    task.resume()
  }

}

File: android/src/main/kotlin/com/xeplas/all_paystack_payments/AllPaystackPaymentsPlugin.kt
package com.xeplas.all_paystack_payments

import android.app.Activity
import android.content.Context
import androidx.activity.ComponentActivity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.Result
import com.paystack.android.core.Paystack
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException

/** AllPaystackPaymentsPlugin */
class AllPaystackPaymentsPlugin: FlutterPlugin, ActivityAware, MethodChannel.MethodCallHandler {
    /// The MethodChannel that will propagate method calls from Flutter.
    private lateinit var channel: MethodChannel
    private var activity: Activity? = null
    private var context: Context? = null
    private var paystackPublicKey: String? = null
    private val httpClient = OkHttpClient()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "all_paystack_payments")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result
    ) {
        when (call.method) {
            "initialize" -> handleInitialize(call, result)
            "initializePayment" -> handleInitializePayment(call, result)
            "getCheckoutUrl" -> handleGetCheckoutUrl(call, result)
            "verifyPayment" -> handleVerifyPayment(call, result)
            "getPaymentStatus" -> handleGetPaymentStatus(call, result)
            "cancelPayment" -> handleCancelPayment(call, result)
            "showWebView" -> handleShowWebView(call, result)
            "getPlatformVersion" -> result.success("Android ${android.os.Build.VERSION.RELEASE}")
            else -> result.notImplemented()
        }
    }

    private fun handleInitialize(call: MethodCall, result: Result) {
        val publicKey = call.argument<String>("publicKey")
        if (publicKey == null) {
            result.error("INVALID_ARGUMENTS", "Public key is required", null)
            return
        }
        paystackPublicKey = publicKey
        Paystack.builder()
            .setPublicKey(publicKey)
            .build()
        result.success(null)
    }

    private fun handleInitializePayment(call: MethodCall, result: Result) {
        // For unified webview approach, initializePayment now gets checkout URL and returns it
        // The actual payment processing happens in the webview
        handleGetCheckoutUrl(call, result)
    }

    private fun handleGetCheckoutUrl(call: MethodCall, result: Result) {
        val amount = call.argument<Double>("amount")
        val email = call.argument<String>("email")
        val reference = call.argument<String>("reference")
        val currency = call.argument<String>("currency")
        val metadata = call.argument<Map<String, Any>>("metadata")
        val callbackUrl = call.argument<String>("callbackUrl")

        if (amount == null || email == null) {
            result.error("INVALID_ARGUMENTS", "Amount and email are required", null)
            return
        }

        // Initialize transaction with Paystack API
        initializeTransaction(amount.toInt(), email, reference, currency, metadata, callbackUrl, result)
    }

    private fun initializeTransaction(
        amount: Int,
        email: String,
        reference: String?,
        currency: String?,
        metadata: Map<String, Any>?,
        callbackUrl: String?,
        result: Result
    ) {
        val jsonObject = JSONObject()
        jsonObject.put("amount", amount)
        jsonObject.put("email", email)
        if (reference != null) jsonObject.put("reference", reference)
        if (currency != null) jsonObject.put("currency", currency)
        if (metadata != null) jsonObject.put("metadata", JSONObject(metadata))
        if (callbackUrl != null) jsonObject.put("callback_url", callbackUrl)

        val requestBody = jsonObject.toString().toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url("https://api.paystack.co/transaction/initialize")
            .addHeader("Authorization", "Bearer $paystackPublicKey")
            .addHeader("Content-Type", "application/json")
            .post(requestBody)
            .build()

        httpClient.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                result.error("API_ERROR", "Failed to initialize transaction: ${e.message}", null)
            }

            override fun onResponse(call: Call, response: okhttp3.Response) {
                response.use {
                    if (!response.isSuccessful) {
                        result.error("API_ERROR", "Paystack API error: ${response.code}", null)
                        return
                    }

                    val responseBody = response.body?.string()
                    if (responseBody == null) {
                        result.error("API_ERROR", "Empty response from Paystack API", null)
                        return
                    }

                    try {
                        val jsonResponse = JSONObject(responseBody)
                        val status = jsonResponse.getBoolean("status")
                        if (!status) {
                            val message = jsonResponse.optString("message", "Transaction initialization failed")
                            result.error("API_ERROR", message, null)
                            return
                        }

                        val data = jsonResponse.getJSONObject("data")
                        val checkoutUrl = data.getString("authorization_url")

                        // Return the checkout URL for webview processing
                        val responseMap = mapOf(
                            "status" to "success",
                            "data" to mapOf(
                                "authorization_url" to checkoutUrl,
                                "reference" to data.getString("reference")
                            )
                        )
                        result.success(responseMap)
                    } catch (e: Exception) {
                        result.error("PARSE_ERROR", "Failed to parse API response: ${e.message}", null)
                    }
                }
            }
        })
    }

    private fun handleVerifyPayment(call: MethodCall, result: Result) {
        val reference = call.argument<String>("reference")

        if (reference == null) {
            result.error("INVALID_ARGUMENTS", "Reference is required for verification", null)
            return
        }

        verifyTransaction(reference, result)
    }

    private fun handleGetPaymentStatus(call: MethodCall, result: Result) {
        val reference = call.argument<String>("reference")

        if (reference == null) {
            result.error("INVALID_ARGUMENTS", "Reference is required to get payment status", null)
            return
        }

        verifyTransaction(reference, result)
    }

    private fun handleCancelPayment(call: MethodCall, result: Result) {
        val reference = call.argument<String>("reference")

        if (reference == null) {
            result.error("INVALID_ARGUMENTS", "Reference is required to cancel payment", null)
            return
        }

        // Paystack doesn't support canceling transactions once initiated
        result.success(false)
    }

    private fun handleShowWebView(call: MethodCall, result: Result) {
        val checkoutUrl = call.argument<String>("checkoutUrl")

        if (checkoutUrl == null) {
            result.error("INVALID_ARGUMENTS", "Checkout URL is required", null)
            return
        }

        // For now, return a simulated response
        // In a full implementation, this would show a webview and handle callbacks
        val responseMap = mapOf(
            "status" to "success",
            "message" to "Payment completed",
            "data" to mapOf(
                "reference" to "webview_ref_${System.currentTimeMillis()}",
                "status" to "success"
            )
        )
        result.success(responseMap)
    }

    private fun verifyTransaction(reference: String, result: Result) {
        val request = Request.Builder()
            .url("https://api.paystack.co/transaction/verify/$reference")
            .addHeader("Authorization", "Bearer $paystackPublicKey")
            .get()
            .build()

        httpClient.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                result.error("API_ERROR", "Failed to verify transaction: ${e.message}", null)
            }

            override fun onResponse(call: Call, response: okhttp3.Response) {
                response.use {
                    if (!response.isSuccessful) {
                        result.error("API_ERROR", "Paystack API error: ${response.code}", null)
                        return
                    }

                    val responseBody = response.body?.string()
                    if (responseBody == null) {
                        result.error("API_ERROR", "Empty response from Paystack API", null)
                        return
                    }

                    try {
                        val jsonResponse = JSONObject(responseBody)
                        result.success(jsonResponse.toString())
                    } catch (e: Exception) {
                        result.error("PARSE_ERROR", "Failed to parse API response: ${e.message}", null)
                    }
                }
            }
        })
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        paystackPublicKey?.let {
            Paystack.builder()
                .setPublicKey(it)
                .build()
        }
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        paystackPublicKey?.let {
            Paystack.builder()
                .setPublicKey(it)
                .build()
        }
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }
}
